File: src/main/java/fi/quanfoxes/phases/LexerPhase.java
package fi.quanfoxes.phases;

import java.util.ArrayList;
import java.util.List;

import fi.quanfoxes.Bundle;
import fi.quanfoxes.Phase;
import fi.quanfoxes.Status;
import fi.quanfoxes.lexer.Lexer;
import fi.quanfoxes.lexer.Token;

public class LexerPhase extends Phase {

    @Override
    public Status execute(Bundle bundle) {
        final String[] contents = bundle.get("input_file_contents", new String[] {});

        if (contents.length == 0) {
            return Status.error("Nothing to tokenize");
        }

        @SuppressWarnings("unchecked")
        final List<Token>[] tokens = new ArrayList[contents.length];

        for (int i = 0; i < contents.length; i++) {
            final int index = i;

            async(() -> {
                String content = contents[index];

                try {
                    tokens[index] = Lexer.getTokens(content);
                }
                catch (Exception e) {
                    return Status.error(e.getMessage());
                }

                return Status.OK;
            });
        }

        bundle.put("input_file_tokens", tokens);

        return Status.OK;
    }
}

File: Zigzag/Phases/LexerPhase.cs
using System.Collections.Generic;
using System;

public class LexerPhase : Phase
{
	public override Status Execute(Bundle bundle)
	{
		string[] contents = bundle.Get("input_file_contents", new string[] { });

		if (contents.Length == 0)
		{
			return Status.Error("Nothing to tokenize");
		}

		List<Token>[] tokens = new List<Token>[contents.Length];

		for (int i = 0; i < contents.Length; i++)
		{
			int index = i;

			Async(() =>
			{
				string content = contents[index];

				try
				{
					tokens[index] = Lexer.GetTokens(content);
				}
				catch (Exception e)
				{
					return Status.Error(e.Message);
				}

				return Status.OK;
			});
		}

		bundle.Put("input_file_tokens", tokens);

		return Status.OK;
	}
}